include/umf/memspace.h

```diff
@@ -0,0 +1,43 @@
+/*
+ *
+ * Copyright (C) 2023 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#ifndef UMF_MEMSPACE_H
+#define UMF_MEMSPACE_H 1
+
+#include <umf/base.h>
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+#include <umf/memspace_policy.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct umf_memspace_t *umf_memspace_handle_t;
+
+// TODO: consider 3rd function that would only return pool ops (as opposed
+// to actual pool instance with a provider). This could be useful for UR,
+// where we could ask UMF for a best pool (e.g. based on policy) but supply
+// a custom provider (L0/CUDA, etc.).
+enum umf_result_t umfPoolCreateFromMemspace(umf_memspace_handle_t memspace,
+                                            umf_memspace_policy_handle_t policy,
+                                            umf_memory_pool_handle_t *pool);
+enum umf_result_t
+umfMemoryProviderCreateFromMemspace(umf_memspace_handle_t memspace,
+                                    umf_memspace_policy_handle_t policy,
+                                    umf_memory_provider_handle_t *provider);
+
+// TODO: iteration, filtering, sorting API
+// can we just define memspace as an array of targets?
```

Comment:
// can we just define memspace as an array of targets?
We can do that only if we're sure our memspaces won't contain any private data. This could be some common properties and data related to filtering or ordering - consider an example where the ordering would be "dynamic" so it would be applied before each allocation request (like order by current free space). If we support dynamic filtering or ordering, it has to be stored somewhere in the memspace priv data.

Reply:
Do we have a use case for dynamic filtering or ordering? I don't really see a benefit of doing this. We would need to apply the filtering/ordering any time user calls iteration API anyway. So why not just apply the filtering/ordering immediately and have an array of always sorted/filtered memory_targets. With that, user can just do:

```
for (size_t i = 0; i < memspaceHandle->size; i++) {
    memspaceHandle->memoryTarget[i]->doSomething;
}
```

Reply:
I agree with Igor that the dynamic allocation policy (if it is allowed) should be handled by the memory provider because the provider is responsible for actual memory allocation from the memory target. Memory provider is instantiated and configured based on input parameters of `umfMemoryProviderCreateFromMemspace` function.
But anyway we need to decide what is encapsulated by each parameter. Form my perspective memspace can encapsulate information about memory location (e.g. memory targets) and policy can define how memory location(s) are used (e.g. local first, interleaving, interleaving size, etc.).

Any thoughts/comments/objections?

Reply:
Agree. However, I think we can have some 'dynamic' metrics as part of the memspace: total available memory in the system, etc. that could be queried. But I don't think making the memspace dynamically sorted based on this information is a good idea. Especially since those metrics can changed right after the memspace has been sorted and we'll have inconsistent state.